File: mobile/lib/services/analytics_service.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';
import 'api_client.dart';

final analyticsServiceProvider = Provider<AnalyticsService>((ref) {
  final apiClient = ref.read(apiClientProvider);
  return AnalyticsService(apiClient);
});

class AnalyticsService {
  final ApiClient _apiClient;

  AnalyticsService(this._apiClient);

  /// Get spending summary for a date range
  Future<AnalyticsSummary> getSummary({
    DateTime? from,
    DateTime? to,
  }) async {
    try {
      final Map<String, dynamic> params = {};
      if (from != null) {
        params['from'] = from.toIso8601String().split('T')[0];
      }
      if (to != null) {
        params['to'] = to.toIso8601String().split('T')[0];
      }

      final response = await _apiClient.get(
        '/api/analytics/summary',
        queryParameters: params,
      );

      if (response.statusCode == 200) {
        return AnalyticsSummary.fromJson(response.data['data']);
      }
      throw Exception('Failed to fetch summary');
    } catch (e) {
      rethrow;
    }
  }

  /// Get time series data for charts
  Future<List<TimeSeriesData>> getTimeSeries({
    String groupBy = 'month',
    DateTime? from,
    DateTime? to,
  }) async {
    try {
      final Map<String, dynamic> params = {
        'groupBy': groupBy,
      };
      if (from != null) {
        params['from'] = from.toIso8601String().split('T')[0];
      }
      if (to != null) {
        params['to'] = to.toIso8601String().split('T')[0];
      }

      final response = await _apiClient.get(
        '/api/analytics/timeseries',
        queryParameters: params,
      );

      if (response.statusCode == 200) {
        final data = response.data['data'] as List;
        return data.map((e) => TimeSeriesData.fromJson(e)).toList();
      }
      throw Exception('Failed to fetch time series');
    } catch (e) {
      rethrow;
    }
  }

  /// Get spending forecast
  Future<SpendingForecast> getForecast() async {
    try {
      final response = await _apiClient.get('/api/analytics/forecast');

      if (response.statusCode == 200) {
        return SpendingForecast.fromJson(response.data['data']);
      }
      throw Exception('Failed to fetch forecast');
    } catch (e) {
      rethrow;
    }
  }

  /// Get top spending categories
  Future<List<CategoryBreakdown>> getTopCategories({int limit = 5}) async {
    try {
      final response = await _apiClient.get(
        '/api/analytics/top-categories',
        queryParameters: {'limit': limit.toString()},
      );

      if (response.statusCode == 200) {
        final data = response.data['data'] as List;
        return data.map((e) => CategoryBreakdown.fromJson(e)).toList();
      }
      throw Exception('Failed to fetch top categories');
    } catch (e) {
      rethrow;
    }
  }

  /// Get period comparison
  Future<PeriodComparison> getPeriodComparison({
    DateTime? from,
    DateTime? to,
  }) async {
    try {
      final Map<String, dynamic> params = {};
      if (from != null) {
        params['from'] = from.toIso8601String().split('T')[0];
      }
      if (to != null) {
        params['to'] = to.toIso8601String().split('T')[0];
      }

      final response = await _apiClient.get(
        '/api/analytics/comparison',
        queryParameters: params,
      );

      if (response.statusCode == 200) {
        return PeriodComparison.fromJson(response.data['data']);
      }
      throw Exception('Failed to fetch comparison');
    } catch (e) {
      rethrow;
    }
  }
}

// Response models
class AnalyticsSummary {
  final double totalIncome;
  final double totalExpense;
  final double netSavings;
  final double savingsRate;
  final int transactionCount;
  final List<CategoryBreakdown> categoryBreakdown;

  AnalyticsSummary({
    required this.totalIncome,
    required this.totalExpense,
    required this.netSavings,
    required this.savingsRate,
    required this.transactionCount,
    required this.categoryBreakdown,
  });

  factory AnalyticsSummary.fromJson(Map<String, dynamic> json) {
    return AnalyticsSummary(
      totalIncome: (json['totalIncome'] ?? 0).toDouble(),
      totalExpense: (json['totalExpense'] ?? 0).toDouble(),
      netSavings: (json['netSavings'] ?? 0).toDouble(),
      savingsRate: (json['savingsRate'] ?? 0).toDouble(),
      transactionCount: json['transactionCount'] ?? 0,
      categoryBreakdown: (json['categoryBreakdown'] as List?)
          ?.map((e) => CategoryBreakdown.fromJson(e))
          .toList() ?? [],
    );
  }
}

class CategoryBreakdown {
  final int categoryId;
  final String categoryName;
  final String? icon;
  final String? color;
  final double total;
  final double percentage;
  final int count;

  CategoryBreakdown({
    required this.categoryId,
    required this.categoryName,
    this.icon,
    this.color,
    required this.total,
    required this.percentage,
    required this.count,
  });

  factory CategoryBreakdown.fromJson(Map<String, dynamic> json) {
    return CategoryBreakdown(
      categoryId: json['categoryId'] ?? 0,
      categoryName: json['categoryName'] ?? '',
      icon: json['icon'],
      color: json['color'],
      total: (json['total'] ?? 0).toDouble(),
      percentage: (json['percentage'] ?? 0).toDouble(),
      count: json['count'] ?? 0,
    );
  }
}

class TimeSeriesData {
  final String label;
  final double totalExpense;
  final double totalIncome;
  final double netSavings;

  TimeSeriesData({
    required this.label,
    required this.totalExpense,
    required this.totalIncome,
    required this.netSavings,
  });

  factory TimeSeriesData.fromJson(Map<String, dynamic> json) {
    return TimeSeriesData(
      label: json['label'] ?? '',
      totalExpense: (json['totalExpense'] ?? 0).toDouble(),
      totalIncome: (json['totalIncome'] ?? 0).toDouble(),
      netSavings: (json['netSavings'] ?? 0).toDouble(),
    );
  }
}

class SpendingForecast {
  final double expectedTotalExpenseNextMonth;
  final List<CategoryForecast> expectedByCategory;
  final int confidence;
  final int basedOnMonths;

  SpendingForecast({
    required this.expectedTotalExpenseNextMonth,
    required this.expectedByCategory,
    required this.confidence,
    required this.basedOnMonths,
  });

  factory SpendingForecast.fromJson(Map<String, dynamic> json) {
    return SpendingForecast(
      expectedTotalExpenseNextMonth:
          (json['expectedTotalExpenseNextMonth'] ?? 0).toDouble(),
      expectedByCategory: (json['expectedByCategory'] as List?)
          ?.map((e) => CategoryForecast.fromJson(e))
          .toList() ?? [],
      confidence: json['confidence'] ?? 0,
      basedOnMonths: json['basedOnMonths'] ?? 0,
    );
  }
}

class CategoryForecast {
  final int categoryId;
  final String categoryName;
  final double expectedAmount;
  final double averageAmount;
  final String trend;

  CategoryForecast({
    required this.categoryId,
    required this.categoryName,
    required this.expectedAmount,
    required this.averageAmount,
    required this.trend,
  });

  factory CategoryForecast.fromJson(Map<String, dynamic> json) {
    return CategoryForecast(
      categoryId: json['categoryId'] ?? 0,
      categoryName: json['categoryName'] ?? '',
      expectedAmount: (json['expectedAmount'] ?? 0).toDouble(),
      averageAmount: (json['averageAmount'] ?? 0).toDouble(),
      trend: json['trend'] ?? 'stable',
    );
  }
}

class PeriodComparison {
  final PeriodData currentPeriod;
  final PeriodData previousPeriod;
  final ComparisonChanges changes;

  PeriodComparison({
    required this.currentPeriod,
    required this.previousPeriod,
    required this.changes,
  });

  factory PeriodComparison.fromJson(Map<String, dynamic> json) {
    return PeriodComparison(
      currentPeriod: PeriodData.fromJson(json['currentPeriod'] ?? {}),
      previousPeriod: PeriodData.fromJson(json['previousPeriod'] ?? {}),
      changes: ComparisonChanges.fromJson(json['changes'] ?? {}),
    );
  }
}

class PeriodData {
  final double income;
  final double expense;
  final double savings;

  PeriodData({
    required this.income,
    required this.expense,
    required this.savings,
  });

  factory PeriodData.fromJson(Map<String, dynamic> json) {
    return PeriodData(
      income: (json['income'] ?? 0).toDouble(),
      expense: (json['expense'] ?? 0).toDouble(),
      savings: (json['savings'] ?? 0).toDouble(),
    );
  }
}

class ComparisonChanges {
  final double incomeChange;
  final double expenseChange;
  final double savingsChange;
  final double incomeChangePercent;
  final double expenseChangePercent;
  final double savingsChangePercent;

  ComparisonChanges({
    required this.incomeChange,
    required this.expenseChange,
    required this.savingsChange,
    required this.incomeChangePercent,
    required this.expenseChangePercent,
    required this.savingsChangePercent,
  });

  factory ComparisonChanges.fromJson(Map<String, dynamic> json) {
    return ComparisonChanges(
      incomeChange: (json['incomeChange'] ?? 0).toDouble(),
      expenseChange: (json['expenseChange'] ?? 0).toDouble(),
      savingsChange: (json['savingsChange'] ?? 0).toDouble(),
      incomeChangePercent: (json['incomeChangePercent'] ?? 0).toDouble(),
      expenseChangePercent: (json['expenseChangePercent'] ?? 0).toDouble(),
      savingsChangePercent: (json['savingsChangePercent'] ?? 0).toDouble(),
    );
  }
}

File: backend/src/services/analytics.service.ts
import { PrismaClient } from '@prisma/client';

const prisma = new PrismaClient();

export interface SummaryResult {
  totalIncome: number;
  totalExpense: number;
  netSavings: number;
  savingsRate: number;
  transactionCount: number;
  categoryBreakdown: Array<{
    categoryId: number;
    categoryName: string;
    icon: string | null;
    color: string | null;
    total: number;
    percentage: number;
    count: number;
  }>;
}

export interface TimeSeriesData {
  label: string;
  totalExpense: number;
  totalIncome: number;
  netSavings: number;
}

export interface CategoryForecast {
  categoryId: number;
  categoryName: string;
  expectedAmount: number;
  averageAmount: number;
  trend: 'increasing' | 'decreasing' | 'stable';
}

export class AnalyticsService {
  /**
   * Get spending summary for a date range
   * GET /api/analytics/summary
   */
  async getSummary(userId: string, from: Date, to: Date): Promise<SummaryResult> {
    const transactions = await prisma.transaction.findMany({
      where: {
        userId,
        postedAt: { gte: from, lte: to },
      },
      include: {
        category: {
          select: {
            id: true,
            name: true,
            icon: true,
            color: true,
          },
        },
      },
    });

    const income = transactions
      .filter((t) => t.type === 'income')
      .reduce((sum, t) => sum + Number(t.amount), 0);

    const expense = transactions
      .filter((t) => t.type === 'expense')
      .reduce((sum, t) => sum + Number(t.amount), 0);

    // Category breakdown for expenses
    const categoryMap = new Map<number, {
      categoryId: number;
      categoryName: string;
      icon: string | null;
      color: string | null;
      total: number;
      count: number;
    }>();

    transactions
      .filter((t) => t.type === 'expense' && t.category)
      .forEach((t) => {
        const catId = t.categoryId!;
        if (!categoryMap.has(catId)) {
          categoryMap.set(catId, {
            categoryId: catId,
            categoryName: t.category!.name,
            icon: t.category!.icon,
            color: t.category!.color,
            total: 0,
            count: 0,
          });
        }
        const cat = categoryMap.get(catId)!;
        cat.total += Number(t.amount);
        cat.count += 1;
      });

    const categoryBreakdown = Array.from(categoryMap.values())
      .map((cat) => ({
        ...cat,
        percentage: expense > 0 ? (cat.total / expense) * 100 : 0,
      }))
      .sort((a, b) => b.total - a.total);

    return {
      totalIncome: income,
      totalExpense: expense,
      netSavings: income - expense,
      savingsRate: income > 0 ? ((income - expense) / income) * 100 : 0,
      transactionCount: transactions.length,
      categoryBreakdown,
    };
  }

  /**
   * Get time series data for charts
   * GET /api/analytics/timeseries
   */
  async getTimeSeries(
    userId: string,
    from: Date,
    to: Date,
    groupBy: 'day' | 'month' | 'year' = 'month'
  ): Promise<TimeSeriesData[]> {
    const transactions = await prisma.transaction.findMany({
      where: {
        userId,
        postedAt: { gte: from, lte: to },
      },
      orderBy: { postedAt: 'asc' },
    });

    const groupedData = new Map<string, { income: number; expense: number }>();

    transactions.forEach((t) => {
      const date = new Date(t.postedAt);
      let label: string;

      switch (groupBy) {
        case 'day':
          label = `${date.getFullYear()}-${String(date.getMonth() + 1).padStart(2, '0')}-${String(date.getDate()).padStart(2, '0')}`;
          break;
        case 'month':
          label = `${date.getFullYear()}-${String(date.getMonth() + 1).padStart(2, '0')}`;
          break;
        case 'year':
          label = `${date.getFullYear()}`;
          break;
      }

      if (!groupedData.has(label)) {
        groupedData.set(label, { income: 0, expense: 0 });
      }

      const data = groupedData.get(label)!;
      if (t.type === 'income') {
        data.income += Number(t.amount);
      } else {
        data.expense += Number(t.amount);
      }
    });

    // Fill in missing dates for continuous chart
    const result: TimeSeriesData[] = [];
    const sortedLabels = Array.from(groupedData.keys()).sort();

    sortedLabels.forEach((label) => {
      const data = groupedData.get(label)!;
      result.push({
        label,
        totalIncome: data.income,
        totalExpense: data.expense,
        netSavings: data.income - data.expense,
      });
    });

    return result;
  }

  /**
   * Get category spending trends
   * GET /api/analytics/category-trends
   */
  async getCategoryTrends(
    userId: string,
    categoryId: number,
    months: number = 6
  ): Promise<{
    categoryName: string;
    monthlyData: Array<{ month: string; amount: number }>;
    averageMonthly: number;
    trend: 'increasing' | 'decreasing' | 'stable';
  }> {
    const startDate = new Date();
    startDate.setMonth(startDate.getMonth() - months);
    startDate.setDate(1);

    const category = await prisma.category.findUnique({
      where: { id: categoryId },
    });

    if (!category) {
      throw new Error('Category not found');
    }

    const transactions = await prisma.transaction.findMany({
      where: {
        userId,
        categoryId,
        type: 'expense',
        postedAt: { gte: startDate },
      },
      orderBy: { postedAt: 'asc' },
    });

    // Group by month
    const monthlyMap = new Map<string, number>();

    transactions.forEach((t) => {
      const date = new Date(t.postedAt);
      const label = `${date.getFullYear()}-${String(date.getMonth() + 1).padStart(2, '0')}`;
      monthlyMap.set(label, (monthlyMap.get(label) || 0) + Number(t.amount));
    });

    const monthlyData = Array.from(monthlyMap.entries())
      .map(([month, amount]) => ({ month, amount }))
      .sort((a, b) => a.month.localeCompare(b.month));

    // Calculate average and trend
    const totalAmount = monthlyData.reduce((sum, m) => sum + m.amount, 0);
    const averageMonthly = monthlyData.length > 0 ? totalAmount / monthlyData.length : 0;

    // Simple trend detection
    let trend: 'increasing' | 'decreasing' | 'stable' = 'stable';
    if (monthlyData.length >= 3) {
      const recentHalf = monthlyData.slice(-Math.ceil(monthlyData.length / 2));
      const olderHalf = monthlyData.slice(0, Math.floor(monthlyData.length / 2));

      const recentAvg = recentHalf.reduce((sum, m) => sum + m.amount, 0) / recentHalf.length;
      const olderAvg = olderHalf.reduce((sum, m) => sum + m.amount, 0) / olderHalf.length;

      const changePercent = olderAvg > 0 ? ((recentAvg - olderAvg) / olderAvg) * 100 : 0;

      if (changePercent > 10) {
        trend = 'increasing';
      } else if (changePercent < -10) {
        trend = 'decreasing';
      }
    }

    return {
      categoryName: category.name,
      monthlyData,
      averageMonthly,
      trend,
    };
  }

  /**
   * Get spending forecast by category
   * GET /api/analytics/forecast
   */
  async getSpendingForecast(userId: string): Promise<{
    expectedTotalExpenseNextMonth: number;
    expectedByCategory: CategoryForecast[];
    confidence: number;
    basedOnMonths: number;
  }> {
    // Get last 6 months of data
    const sixMonthsAgo = new Date();
    sixMonthsAgo.setMonth(sixMonthsAgo.getMonth() - 6);
    sixMonthsAgo.setDate(1);

    const transactions = await prisma.transaction.findMany({
      where: {
        userId,
        type: 'expense',
        postedAt: { gte: sixMonthsAgo },
      },
      include: {
        category: true,
      },
    });

    // Group by month and category
    const categoryMonthlyMap = new Map<number, Map<string, number>>();
    const categoryNames = new Map<number, string>();

    transactions.forEach((t) => {
      if (!t.categoryId || !t.category) return;

      if (!categoryMonthlyMap.has(t.categoryId)) {
        categoryMonthlyMap.set(t.categoryId, new Map());
        categoryNames.set(t.categoryId, t.category.name);
      }

      const date = new Date(t.postedAt);
      const monthKey = `${date.getFullYear()}-${date.getMonth() + 1}`;
      const monthMap = categoryMonthlyMap.get(t.categoryId)!;
      monthMap.set(monthKey, (monthMap.get(monthKey) || 0) + Number(t.amount));
    });

    // Calculate forecast for each category
    const expectedByCategory: CategoryForecast[] = [];
    let totalExpected = 0;

    categoryMonthlyMap.forEach((monthMap, categoryId) => {
      const values = Array.from(monthMap.values());
      
      if (values.length < 2) return;

      // Simple moving average with trend adjustment
      const average = values.reduce((sum, v) => sum + v, 0) / values.length;
      
      // Check recent trend
      const recentValues = values.slice(-3);
      const recentAvg = recentValues.reduce((sum, v) => sum + v, 0) / recentValues.length;
      
      const trendFactor = average > 0 ? recentAvg / average : 1;
      const adjustedExpected = average * Math.min(Math.max(trendFactor, 0.8), 1.2);

      let trend: 'increasing' | 'decreasing' | 'stable' = 'stable';
      if (trendFactor > 1.1) trend = 'increasing';
      else if (trendFactor < 0.9) trend = 'decreasing';

      expectedByCategory.push({
        categoryId,
        categoryName: categoryNames.get(categoryId) || 'Unknown',
        expectedAmount: Math.round(adjustedExpected),
        averageAmount: Math.round(average),
        trend,
      });

      totalExpected += adjustedExpected;
    });

    // Sort by expected amount
    expectedByCategory.sort((a, b) => b.expectedAmount - a.expectedAmount);

    // Calculate confidence based on data quantity
    const uniqueMonths = new Set<string>();
    transactions.forEach((t) => {
      const date = new Date(t.postedAt);
      uniqueMonths.add(`${date.getFullYear()}-${date.getMonth()}`);
    });

    const monthCount = uniqueMonths.size;
    const confidence = Math.min(monthCount / 6 * 100, 100);

    return {
      expectedTotalExpenseNextMonth: Math.round(totalExpected),
      expectedByCategory,
      confidence: Math.round(confidence),
      basedOnMonths: monthCount,
    };
  }

  /**
   * Get top spending categories for current month
   */
  async getTopCategories(userId: string, limit: number = 5): Promise<Array<{
    categoryId: number;
    categoryName: string;
    icon: string | null;
    color: string | null;
    total: number;
    percentage: number;
  }>> {
    const now = new Date();
    const monthStart = new Date(now.getFullYear(), now.getMonth(), 1);
    const monthEnd = new Date(now.getFullYear(), now.getMonth() + 1, 0, 23, 59, 59);

    const summary = await this.getSummary(userId, monthStart, monthEnd);
    
    return summary.categoryBreakdown.slice(0, limit);
  }

  /**
   * Get spending comparison with previous period
   */
  async getPeriodComparison(
    userId: string,
    currentFrom: Date,
    currentTo: Date
  ): Promise<{
    currentPeriod: { income: number; expense: number; savings: number };
    previousPeriod: { income: number; expense: number; savings: number };
    changes: {
      incomeChange: number;
      expenseChange: number;
      savingsChange: number;
      incomeChangePercent: number;
      expenseChangePercent: number;
      savingsChangePercent: number;
    };
  }> {
    // Calculate previous period dates (same duration before current period)
    const duration = currentTo.getTime() - currentFrom.getTime();
    const previousFrom = new Date(currentFrom.getTime() - duration);
    const previousTo = new Date(currentFrom.getTime() - 1);

    const [currentSummary, previousSummary] = await Promise.all([
      this.getSummary(userId, currentFrom, currentTo),
      this.getSummary(userId, previousFrom, previousTo),
    ]);

    const calculateChangePercent = (current: number, previous: number): number => {
      if (previous === 0) return current > 0 ? 100 : 0;
      return ((current - previous) / previous) * 100;
    };

    return {
      currentPeriod: {
        income: currentSummary.totalIncome,
        expense: currentSummary.totalExpense,
        savings: currentSummary.netSavings,
      },
      previousPeriod: {
        income: previousSummary.totalIncome,
        expense: previousSummary.totalExpense,
        savings: previousSummary.netSavings,
      },
      changes: {
        incomeChange: currentSummary.totalIncome - previousSummary.totalIncome,
        expenseChange: currentSummary.totalExpense - previousSummary.totalExpense,
        savingsChange: currentSummary.netSavings - previousSummary.netSavings,
        incomeChangePercent: calculateChangePercent(
          currentSummary.totalIncome,
          previousSummary.totalIncome
        ),
        expenseChangePercent: calculateChangePercent(
          currentSummary.totalExpense,
          previousSummary.totalExpense
        ),
        savingsChangePercent: calculateChangePercent(
          currentSummary.netSavings,
          previousSummary.netSavings
        ),
      },
    };
  }
}

